orttraining/orttraining/core/graph/training_op_defs.cc

```diff
@@ -1265,6 +1263,90 @@ void RegisterTrainingOpSchemas() {
+      .Input(4, "momentums_1", "Sequence of exponentially averaged historical gradients.", "S_MOMENT")
+      .Input(5, "momentums_2", "Sequence of exponentially averaged historical squared gradients.", "S_MOMENT")
+      .Output(0, "updated_flag", "Whether gradient is applied or not.", "T2")
+      .Output(1, "updated_weights", "Sequence of weights after optimize.", "S_WEIGHT", OpSchema::Optional)
+      .Output(2, "updated_momentums_1", "Sequence of momentum_1 after optimize.", "S_MOMENT", OpSchema::Optional)
+      .Output(3, "updated_momentums_2", "Sequence of momentum_2 after optimize.", "S_MOMENT", OpSchema::Optional)
+      .Attr(
+          "alpha",
+          "Coefficient of previously accumulated gradient in running average.",
+          AttributeProto::FLOAT,
+          0.9f)
+      .Attr(
+          "beta",
+          "Coefficient of previously accumulated squared-gradient in running average.",
+          AttributeProto::FLOAT,
+          0.999f)
+      .Attr(
+          "epsilon",
+          "Small scalar to avoid dividing by zero.",
+          AttributeProto::FLOAT,
+          1e-8f)
+      .Attr(
+          "weight_decay",
+          "weight decay coefficient.",
+          AttributeProto::FLOAT,
+          1e-2f)
+      .Attr(
+          "correct_bias",
+          "Whether or not to correct bias, enabled by default.",
+          AttributeProto::INT,
+          static_cast<int64_t>(1))
+      .Attr(
+          "adam_mode",
+          "Modes for applying bias correction and weight decay (default 0) "
+          "0 : Weight decay is applied before weight is updated."
+          "  compuatation aligned with Torch AdamW. In this mode, "
+          "  correct_bias should be set 1 to keep aligned with PyTorch."
+          "1 : Weight decay is applied after weight is updated."
+          "    Computation is aligned with Huggingface AdamW.",
+          AttributeProto::INT,
+          static_cast<int64_t>(0))
+      .TypeConstraint(
+          "T1",
+          {"tensor(float)"},
+          "Constrain learning rate to float")
+      .TypeConstraint(
+          "T2",
+          {"int64"},
+          "Constrain step count to 64-bit integer")
+      .TypeConstraint(
+          "S_WEIGHT",
+          {"seq(tensor(float16))", "seq(tensor(float))", "seq(tensor(double))"},
+          "Constrain weights' types.")
+      .TypeConstraint(
+          "S_GRAD",
+          {"seq(tensor(float16))", "seq(tensor(float))", "seq(tensor(double))"},
+          "Constrain gradients' types.")
+      .TypeConstraint(
+          "S_MOMENT",
+          {"seq(tensor(float16))", "seq(tensor(float))", "seq(tensor(double))"},
+          "Constrain momentums' types.")
+      .TypeAndShapeInferenceFunction([](InferenceContext& ctx) {
+        if (ctx.getNumOutputs() >= 2) {
+          propagateElemTypeFromInputToOutput(ctx, 2, 1);
+          propagateShapeFromInputToOutput(ctx, 2, 1);
```

Comment:
since you are making an assumption that input shape exists, you may want to explicitly enforce it before line 1335. Something like if (!hasNInputShapes(ctx, 5)) { throw error;}

Reply:
This is more complex than I expected. simply check hasNInputShapes(ctx, 5) and throw, will fail our test cases. 

The reason is, Seq<Tensor>'s shape are inferred this way: 

> // Example 1: same rank, different dimensions
> //    input1 shape: (2, 3, 4, 'x')
> //    input2 shape: (2, 'y', 5, 'x')
> //    output shape: (2, None, None, 'x')
> // Example 2: different rank
> //    input1 shape: (2, 3, 4, 'x')
> //    input2 shape: (2, 3, 4)
> //    output shape: None

For our case, weights having different shapes/ranks, so the seq<tensor> of inputs are empty. 

So we need tolerant some input shape empty, cannot throw directly.